### PR TITLE
Fix bug where binding data is expected to overwrite the related resource name

### LINKED ISSUE #475: Service binding global and service envVarPrefix doesn't behave as expected
According to our latest test, there are some features broken after #407 is merged.

Below is the SBR and the environment result.

```
cat <<EOF | kubectl apply -f -
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest 
metadata:
  name: sbr5
spec:
  envVarPrefix: globalprefix
  applicationSelector:
    resourceRef: mytranslator-web-vcap5
    group: apps
    version: v1
    resource: deployments 
  backingServiceSelectors:
    - group: ibmcloud.ibm.com
      version: v1alpha1
      kind: Binding
      resourceRef: thetranslator5
      envVarPrefix: prefix1
  customEnvVar:
    - name: ALL_SECRET
      value: '{{json .}}'
EOF
```

Env var result:

```
GLOBALPREFIX_PREFIX1_BINDING_SECRET_STATUS_SECRETNAME_IAM_ROLE_CRN=crn:v1:bluemix:public:iam::::serviceRole:Manager
GLOBALPREFIX_PREFIX1_BINDING_SECRET_STATUS_SECRETNAME_IAM_APIKEY_DESCRIPTION=Auto-generated for key 0aef56aa-fbb2-4dd7-8bc7-972302e14573
GLOBALPREFIX_PREFIX1_BINDING_SECRET_STATUS_SECRETNAME_APIKEY=xxxxx
globalprefix_ALL_SECRET={"v1alpha1":{"ibmcloud.ibm.com":{"Binding":{"thetranslator5":{"apiVersion":"ibmcloud.ibm.com/v1alpha1","kind":"Binding","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"ibmcloud.ibm.com/v1alpha1\",\"kind\":\"Binding\",\"metadata\":{\"annotations\":{},\"name\":\"thetranslator5\",\"namespace\":\"default\"},\"spec\":{\"serviceName\":\"thetranslator5\"}}\n","service-binding-operator.apps.openshift.io/binding-name":"sbr5","service-binding-operator.apps.openshift.io/binding-namespace":"default"},"creationTimestamp":"2020-05-14T08:55:22Z","finalizers":["binding.ibmcloud.ibm.com"],"generation":1,"name":"thetranslator5","namespace":"default","ownerReferences":[{"apiVersion":"ibmcloud.ibm.com/v1alpha1","blockOwnerDeletion":true,"controller":true,"kind":"Service","name":"thetranslator5","uid":"a6adaa6b-94ae-421f-8698-1e9ec2b48e64"}],"resourceVersion":"4439896","selfLink":"/apis/ibmcloud.ibm.com/v1alpha1/namespaces/default/bindings/thetranslator5","uid":"e07c5c19-25d8-4016-add9-213366b4259c"},"spec":{"serviceName":"thetranslator5"},"status":{"instanceId":"crn:v1:bluemix:public:language-translator:us-south:a/ef6a34810cbcd892507d3ebe01e3d95a:78abe198-feb8-41c9-b293-38b21beb8973::","keyInstanceId":"crn:v1:bluemix:public:language-translator:us-south:a/ef6a34810cbcd892507d3ebe01e3d95a:78abe198-feb8-41c9-b293-38b21beb8973:resource-key:0aef56aa-fbb2-4dd7-8bc7-972302e14573","message":"Online","secretName":"thetranslator5","state":"Online"}}}},"ibmcloud_ibm_com":{"Binding":{"thetranslator5":{"apiVersion":"ibmcloud.ibm.com/v1alpha1","kind":"Binding","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"ibmcloud.ibm.com/v1alpha1\",\"kind\":\"Binding\",\"metadata\":{\"annotations\":{},\"name\":\"thetranslator5\",\"namespace\":\"default\"},\"spec\":{\"serviceName\":\"thetranslator5\"}}\n","service-binding-operator.apps.openshift.io/binding-name":"sbr5","service-binding-operator.apps.openshift.io/binding-namespace":"default"},"creationTimestamp":"2020-05-14T08:55:22Z","finalizers":["binding.ibmcloud.ibm.com"],"generation":1,"name":"thetranslator5","namespace":"default","ownerReferences":[{"apiVersion":"ibmcloud.ibm.com/v1alpha1","blockOwnerDeletion":true,"controller":true,"kind":"Service","name":"thetranslator5","uid":"a6adaa6b-94ae-421f-8698-1e9ec2b48e64"}],"resourceVersion":"4439896","selfLink":"/apis/ibmcloud.ibm.com/v1alpha1/namespaces/default/bindings/thetranslator5","uid":"e07c5c19-25d8-4016-add9-213366b4259c"},"spec":{"serviceName":"thetranslator5"},"status":{"instanceId":"crn:v1:bluemix:public:language-translator:us-south:a/ef6a34810cbcd892507d3ebe01e3d95a:78abe198-feb8-41c9-b293-38b21beb8973::","keyInstanceId":"crn:v1:bluemix:public:language-translator:us-south:a/ef6a34810cbcd892507d3ebe01e3d95a:78abe198-feb8-41c9-b293-38b21beb8973:resource-key:0aef56aa-fbb2-4dd7-8bc7-972302e14573","message":"Online","secretName":"thetranslator5","state":"Online"}}}}}}
GLOBALPREFIX_PREFIX1_BINDING_SECRET_STATUS_SECRETNAME_IAM_SERVICEID_CRN=crn:v1:bluemix:public:iam-identity::a/ef6a34810cbcd892507d3ebe01e3d95a::serviceid:ServiceId-71e18dc7-1d02-4704-b167-be29f4172044
GLOBALPREFIX_PREFIX1_BINDING_SECRET_STATUS_SECRETNAME_IAM_APIKEY_NAME=thetranslator5
GLOBALPREFIX_PREFIX1_BINDING_SECRET_STATUS_SECRETNAME_URL=https://api.us-south.language-translator.watson.cloud.ibm.com/instances/78abe198-feb8-41c9-b293-38b21beb8973

```
There are 3 issues I have found:

1 There are prefixes `BINDING_SECRET_STATUS` for all env vars, but previously we do not have them.
2 With the custom env var `ALL_SECRET`, there is only secretName in the result content, but previously the content of the secret should be in the content.

### PR DESCRIPTION
**Pull request by isutton**
### Motivation

A bug has been introduced in #407, changing the existing behavior where the data collected from the resource indicated by the annotation (either a `corev1.ConfigMap` or `corev1.Secret`) should also replace the field indicating the resource name with the collected data.

Let's take a look in the following example resource:

```yaml
metadata:
  annotations:
    servicebindingoperator.redhat.io/status.secretName: binding:env:object:secret
status:
  secretName: the-secret
```

Before #407, when `status.secretName` was used in a custom environment variable template, it would contain the whole `data` section of the related Secret, as the following excerpt from #475:

```
"status": {
  "secretName": {
    "apikey": "XXXX",
    "iam_apikey_description": "Auto-generated for key 0aef56aa-fbb2-4dd7-8bc7-972302e14573",
    "iam_apikey_name": "thetranslator5",
    "iam_role_crn": "crn:v1:bluemix:public:iam::::serviceRole:Manager",
    "iam_serviceid_crn": "crn:v1:bluemix:public:iam-identity::a/ef6a34810cbcd892507d3ebe01e3d95a::serviceid:ServiceId-71e18dc7-1d02-4704-b167-be29f4172044",
    "url": "https://api.us-south.language-translator.watson.cloud.ibm.com/instances/78abe198-feb8-41c9-b293-38b21beb8973"
  },
}
```

And after #407, the following results were obtained instead:

```
"status": {
  "secretName": "thetranslator5",
}
```

Closes #475 

### Changes

The contents of a related resource are collected, grouped and made available to the custom environment variable template engine; as an example, if the field path `.status.dbCredentials` represents a resource name, it is expected the related resource contents (more specifically the `.data` field) to be available under the same `.status.dbCredentials` path when composing a custom environment variable template in a Service Binding Request.

Below there is an example of what the input object (the service resource that contains the reference to the related resource) and the output object (the resource that is available to the custom environment template context):

```go
// inputObj is whatever object containing both annotations (either directly in it or through its 
// CRD or CSV.
inputObj := map[string]interface{}{
    "metadata": map[string]interface{}{
        // annotations create the relationship of .status.secretName to a secret
        "annotations": map[string]string{....},
    },
    "status": map[string]interface{}{
        // secretName contains the name of a resource living in the same namespace as the 
        // Service Binding Request.
        "secretName": "thetranslator5",
    },
}

// outputObj is inputPath with '.status.dbCredentials' 
outputObj := map[string]interface{}{
    "status": map[string]interface{}{
        "secretName": map[string]interface{}{
          "apikey": "XXXX",
          "iam_apikey_description": "Auto-generated for key 0aef56aa-fbb2-4dd7-8bc7-972302e14573",
          "iam_apikey_name": "thetranslator5",
          "iam_role_crn": "crn:v1:bluemix:public:iam::::serviceRole:Manager",
          "iam_serviceid_crn": "crn:v1:bluemix:public:iam-identity::a/ef6a34810cbcd892507d3ebe01e3d95a::serviceid:ServiceId-71e18dc7-1d02-4704-b167-be29f4172044",
          "url": "https://api.us-south.language-translator.watson.cloud.ibm.com/instances/78abe198-feb8-41c9-b293-38b21beb8973",
        },
    },
}
```

For further more details refer the CONTRIBUTING.md

### CODE REVIEW

**Comment by sbose78** [comment]
Could you outline the user-facing scenario with an example please?

**Comment by sbose78** [comment]
Could you update the description with an example of what's broken, that's being fixed?

**Comment by isutton** [comment]
> Could you update the description with an example of what's broken, that's being fixed?

@sbose78 updated.

**Comment by sbose78** [comment]
@isutton Is this an issue with consumption in custom environment variables?

**Comment by qibobo** [comment]
@isutton 
There are conflicts and needs rebase.

File: pkg/controller/servicebindingrequest/annotations/annotations.go [hunk 1]
```diff
@@ -40,6 +40,8 @@ type result struct {
 	Type bindingType
 	// Path is the nested location the collected data can be found in the Data field.
 	Path string
+	// RawData is...
```
Comment (Avni-Sharma) [inline pkg/controller/servicebindingrequest/annotations/annotations.go:43]: maybe the description was missed here

Comment (isutton) [reply pkg/controller/servicebindingrequest/annotations/annotations.go]: Solved in 70393d9.

**Comment by qibobo** [comment]
@isutton 
Will this PR be merged soon? PR521 needs rebase after this PR is merged.

**Pull request merged**